### PR TITLE
LVRUBY-94: Add contact us form

### PR DESCRIPTION
**Pull request by kmandziy**
## JIRA ticket

https://ssu-jira.softserveinc.com/browse/LVRUBY-94

## Code reviewers

_Please list your code reviewers here._

## Summary of issue

Add page contact us

## Summary of change

Added page contact us
Created mailer for sending message to dark-side email

## Review checklist

- [x] Rubocop doesn't raise any warnings and offenses
- [x] New features are covered with new specs
- [x] All specs pass
Screenshot from 2019-05-07 03-16-59

### CODE REVIEW

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

File: app/mailers/messages_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class MessagesMailer < ActionMailer::Base
+  def contact_us_email(message)
+    @message = message
+    mail(from: @message.email, to: 'soft.darkside@gmail.com',
+    subject: 'Contact us')
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/mailers/messages_mailer.rb]: Changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  margin: 0 auto;
+  position: relative;
+  width: 100%;
+}
+
+.contact {
+  background: $contact-background;
+  padding: 25px;
+
+  input[type='submit'] {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    border: 0;
+    color: $white;
+    cursor: pointer;
+    font-size: 15px;
+    margin: 0 0 5px;
+    padding: 10px;
+    width: 100%;
+  }
+
+  input[type='submit']:hover {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+fieldset {
+  margin: 0 0 10px;
+  min-width: 100%;
+  padding: 0;
+  width: 100%;
+}
+
+.contact-input, .contact-textarea {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/base/_variables.scss [hunk 5]
```diff
@@ -1,3 +1,7 @@
 $banner-color: #f4f5f7;
 $black: #000000;
+$contact-background: #f9f9f9;
 $main-font: 'Montserrat';
+$submit-color: #0CF;
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/base/_variables.scss]: changed

File: app/assets/stylesheets/base/_variables.scss [hunk 6]
```diff
@@ -1,3 +1,7 @@
...
+$submit-hover-color: #09C;
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/base/_variables.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+fieldset {
+  margin: 0 0 10px;
+  min-width: 100%;
+  padding: 0;
+  width: 100%;
+}
+
+#contact input, #contact textarea {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 8]
```diff
@@ -0,0 +1,47 @@
+.contact-container {
+  margin: 0 auto;
+  position: relative;
+  width: 100%;
+}
+
+#contact {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+fieldset {
+  margin: 0 0 10px;
+  min-width: 100%;
+  padding: 0;
+  width: 100%;
+}
+
+#contact input, #contact textarea {
+  background: $white;
+  border: 1px solid #CCC;
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  margin: 0 0 5px;
+  padding: 10px;
+  width: 100%;
+}
+
+#contact textarea {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  height: 100px;
+  max-width: 100%;
+  resize: none;
+}
+
+#contact input[type='submit'] {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  background: $submit-color;
+  border: none;
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

File: app/assets/stylesheets/pages/contact_us.scss [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  color: $white;
+  cursor: pointer;
+  font-size: 15px;
+  margin: 0 0 5px;
+  padding: 10px;
+  width: 100%;
+}
+
+#contact input[type='submit']:hover {
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/contact_us.scss]: changed

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.4%** covered at **2.59** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.43%** covered at **2.64** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.43%** covered at **2.64** hits/line

File: spec/models/message_spec.rb [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      end
+    end
+
+    context 'when author name is empty' do
+      it 'is not valid' do
+        message.author_name = ' '
+        expect(message).not_to be_valid
+      end
+    end
+  end
```
Comment (olenaniemova) [inline spec/models/message_spec.rb]: Add positive cases

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.43%** covered at **2.64** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.43%** covered at **2.64** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.43%** covered at **2.64** hits/line

File: app/assets/stylesheets/pages/contact_us.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (omaik) [reply app/assets/stylesheets/pages/contact_us.scss]: You write "changed" but seems that it isn't. Could you please explain?

File: app/controllers/messages_controller.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class MessagesController < ApplicationController
+  def contact_us
+    @message = Message.new
+  end
+
+  def create
+    @message = Message.new(message_params)
+    if @message.valid?
```
Comment (omaik) [inline app/controllers/messages_controller.rb]: is it expected that you do not save the message in the database? 

In this case, we probably should not create the table in the database and use only ActiveModel to perform validations. https://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html

File: app/controllers/messages_controller.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+      MessagesMailer.contact_us_email(@message).deliver_now
+      redirect_to contact_us_path, notice: 'Your email was succesfully sent'
```
Comment (omaik) [inline app/controllers/messages_controller.rb]: please use I18n

File: spec/controllers/messages_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  describe '#contact_us' do
+    it 'renders contact_us page' do
+      get :contact_us
+      expect(response).to render_template :contact_us
+    end
+  end
+
+  describe '#create' do
+    context 'when message params is right' do
+      it 'renders flash notice' do
```
Comment (omaik) [inline spec/controllers/messages_controller_spec.rb]: please add one more test checking that email was sent

File: spec/controllers/messages_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        post :create, params: { message: params }
+        expect(flash[:notice]).to be_present
+      end
+    end
+
+    context 'when message params is not right' do
+      it 'renders contact us page' do
```
Comment (omaik) [inline spec/controllers/messages_controller_spec.rb]: please add one more test about email wasn't sent

File: spec/mailers/messages_spec.rb [hunk 19]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
```
Comment (omaik) [inline spec/mailers/messages_spec.rb]: the file name does not correspond to the file being tested

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

File: app/views/application/contact_us.html.haml [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+          = t('contacts.information.telephone')
+        +38 000 270 000
+      %p
+        %b
+          = t('contacts.information.address')
+        8, Sample St.
+    .col-lg-7
+      .contact-container
+        = form_for @message, url: contact_us_url, html: { class: 'contact' } do |form|
+          %h4 Contact us today, and get reply with in 24 hours!
```
Comment (vikabelous) [inline app/views/application/contact_us.html.haml]: i18n

File: spec/mailers/messages_mailer_spec.rb [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+    let(:message) { build(:message) }
+    let(:mail) { MessagesMailer.contact_us_email(message) }
+
+    context 'when user submitted message form' do
+      it 'sends mail from user' do
+        expect(mail.from).to eq([message.email])
+      end
+
+      it 'renders the mail body' do
+        expect(mail.body.encoded).to match(message.email)
```
Comment (vikabelous) [inline spec/mailers/messages_mailer_spec.rb]: Expect body to match email?

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

File: spec/controllers/messages_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+      expect(response).to render_template :contact_us
+    end
+  end
+
+  describe '#create' do
+    context 'when message params is right' do
+      it 'sends message' do
+        deliveries_before = ActionMailer::Base.deliveries.count
+        post :create, params: { message: params }
+        expect(ActionMailer::Base.deliveries.count).to eq deliveries_before + 1
```
Comment (omaik) [inline spec/controllers/messages_controller_spec.rb]: please use this instead - https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/change-matcher

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.49%** covered at **2.61** hits/line

File: app/controllers/messages_controller.rb [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+
+class MessagesController < ApplicationController
+  def contact_us
+    @message = Message.new
+  end
+
+  def create
+    @message = Message.new(message_params)
+    if @message.valid?
+      MessagesMailer.contact_us_email(@message).deliver_now
```
Comment (omaik) [inline app/controllers/messages_controller.rb]: Please use background sending instead

**Pull request merged**